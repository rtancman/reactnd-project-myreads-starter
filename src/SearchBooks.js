import React, { Component } from 'react';
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import Book from './Book'


class SearchBooks extends Component {
  static propTypes = {
    query: PropTypes.string,
    books: PropTypes.array.isRequired,
    onSearch: PropTypes.func.isRequired,
    onMoveShelf: PropTypes.func.isRequired
  }

  state = {
    query: this.props.query || ''
  }

  updateQuery = (query) => {
    this.setState(
      { query: query.trim() },
      () => this.props.onSearch(this.state.query)
    )
  }

  componentDidMount() {
    this.props.onSearch(this.state.query)
  }

  render() {
    const { books, onMoveShelf, onSearch } = this.props
    const { query } = this.state
    return (
      <div className="search-books">
        <div className="search-books-bar">
          <Link to="/" className="close-search">Close</Link>
          <div className="search-books-input-wrapper">
            {/*
              NOTES: The search from BooksAPI is limited to a particular set of search terms.
              You can find these search terms here:
              https://github.com/udacity/reactnd-project-myreads-starter/blob/master/SEARCH_TERMS.md

              However, remember that the BooksAPI.search method DOES search by title or author. So, don't worry if
              you don't find a specific author or title. Every search is limited by search terms.
            */}
            <input
              type="text"
              value={query}
              onChange={(event) => this.updateQuery(event.target.value)}
              placeholder="Search by title or author"
            />
          </div>
        </div>
        <div className="search-books-results">
          <ol className="books-grid">
           {books.map((book) => (
              <li key={book.id}>
                <Book
                  book={book}
                  onMoveShelf={onMoveShelf}
                />
              </li>
            ))}
          </ol>
        </div>
      </div>
    )
  }
}

export default SearchBooks

