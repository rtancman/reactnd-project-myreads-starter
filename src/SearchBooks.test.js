import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom'
import SearchBooks from './SearchBooks'
import renderer from 'react-test-renderer';


const books = [
  {
    'title': 'The Linux Command Line',
    'subtitle': 'A Complete Introduction',
    'authors': [
        'William E. Shotts, Jr.'
    ],
    'imageLinks': {
        'smallThumbnail': 'http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api',
        'thumbnail': 'http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api'
    },
    'id': 'nggnmAEACAAJ',
    'shelf': 'currentlyReading'
  },
  {
    'title': 'The Linux',
    'subtitle': 'Introduction',
    'authors': [
        'Shotts Jr.'
    ],
    'id': 'abcenggnmAEACAAJ',
    'shelf': 'currentlyReading'
  }
]
const onSearch = jest.fn()
const onMoveShelf = jest.fn()

describe('SearchBooks', () => {
  describe('when components renders', () => {
    it('with an empty list books', () => {
      const component = renderer.create(
        <MemoryRouter>
          <SearchBooks
            books={[]}
            onSearch={onSearch}
            onMoveShelf={onMoveShelf}
          />
        </MemoryRouter>
      )
      let tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('with a list books', () => {
      const component = renderer.create(
        <MemoryRouter>
          <SearchBooks
            books={books}
            onSearch={onSearch}
            onMoveShelf={onMoveShelf}
          />
        </MemoryRouter>
      )
      let tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when user search books', () => {
    it('change input value', () => {
      const onMoveShelf = jest.fn();
      const event = {
        target: { value: 'Linux' }
      }
      const wrapper = mount(
        <MemoryRouter>
          <SearchBooks
            query=''
            books={[]}
            onSearch={onSearch}
            onMoveShelf={onMoveShelf}
          />
        </MemoryRouter>
      )
      wrapper.find('input').simulate('change', event);
      expect(onSearch).toBeCalled()
      //expect(wrapper.state().query).toEqual('Linux')
    })
  })
})
