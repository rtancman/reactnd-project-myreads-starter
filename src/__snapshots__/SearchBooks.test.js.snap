// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBooks when components renders with a list books 1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <a
      className="close-search"
      href="/"
      onClick={[Function]}
    >
      Close
    </a>
    <div
      className="search-books-input-wrapper"
    >
      <input
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="search-books-results"
  >
    <ol
      className="books-grid"
    >
      <li>
        <div
          className="book"
        >
          <div
            className="book-top"
          >
            <div
              className="book-cover"
              style={
                Object {
                  "backgroundImage": "url(\\"http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api\\")",
                  "height": 193,
                  "width": 128,
                }
              }
            />
            <div
              className="book-shelf-changer"
            >
              <select
                onChange={[Function]}
                value="currentlyReading"
              >
                <option
                  disabled={true}
                  value="move"
                >
                  Move to...
                </option>
                <option
                  defaultValue={true}
                  value="currentlyReading"
                >
                  Currently Reading
                </option>
                <option
                  defaultValue={false}
                  value="wantToRead"
                >
                  Want to Read
                </option>
                <option
                  defaultValue={false}
                  value="read"
                >
                  Read
                </option>
                <option
                  defaultValue={false}
                  value="none"
                >
                  None
                </option>
              </select>
            </div>
          </div>
          <div
            className="book-title"
          >
            The Linux Command Line
          </div>
          <div
            className="book-authors"
          >
            William E. Shotts, Jr.
          </div>
        </div>
      </li>
      <li>
        <div
          className="book"
        >
          <div
            className="book-top"
          >
            <div
              className="book-shelf-changer"
            >
              <select
                onChange={[Function]}
                value="currentlyReading"
              >
                <option
                  disabled={true}
                  value="move"
                >
                  Move to...
                </option>
                <option
                  defaultValue={true}
                  value="currentlyReading"
                >
                  Currently Reading
                </option>
                <option
                  defaultValue={false}
                  value="wantToRead"
                >
                  Want to Read
                </option>
                <option
                  defaultValue={false}
                  value="read"
                >
                  Read
                </option>
                <option
                  defaultValue={false}
                  value="none"
                >
                  None
                </option>
              </select>
            </div>
          </div>
          <div
            className="book-title"
          >
            The Linux
          </div>
          <div
            className="book-authors"
          >
            Shotts Jr.
          </div>
        </div>
      </li>
    </ol>
  </div>
</div>
`;

exports[`SearchBooks when components renders with an empty list books 1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <a
      className="close-search"
      href="/"
      onClick={[Function]}
    >
      Close
    </a>
    <div
      className="search-books-input-wrapper"
    >
      <input
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="search-books-results"
  >
    <ol
      className="books-grid"
    />
  </div>
</div>
`;
